lib/client/conntest/kube.go

```diff
@@ -0,0 +1,273 @@
+		return connDiag, trace.Wrap(err)
+	}
+
+	if kubeErr != nil && strings.Contains(kubeErr.ErrStatus.Message, "has no assigned groups or users") {
+		message := `User-associated roles do not configure "kubernetes_groups" or "kubernetes_users". Make sure that at least one is configured for the user.`
+		traceType := types.ConnectionDiagnosticTrace_RBAC_PRINCIPAL
+
+		connDiag, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, actionErr)
+		return connDiag, trace.Wrap(err)
+	}
+
+	message := "User-associated roles define valid Kubernetes principals."
+	traceType := types.ConnectionDiagnosticTrace_RBAC_PRINCIPAL
+
+	_, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if kubeErr != nil && (strings.Contains(kubeErr.ErrStatus.Message, "not found") || strings.Contains(kubeErr.ErrStatus.Message, "[00] access denied")) {
+		message := "You are not authorized to access this Kubernetes Cluster. Ensure your role grants access by adding it to the 'kubernetes_labels' property."
+		traceType := types.ConnectionDiagnosticTrace_RBAC_KUBE
+		connDiag, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, actionErr)
+		return connDiag, trace.Wrap(err)
+	}
+
+	if kubeErr != nil && strings.Contains(kubeErr.ErrStatus.Message, "cannot list resource \"pods\"") {
+		message := fmt.Sprintf("You are not allowed to list pods in the %q namespace. "+
+			"Make sure your \"kubernetes_groups\" or \"kubernetes_users\" exist in the cluster and grant you access to list pods.", namespace)
+		traceType := types.ConnectionDiagnosticTrace_KUBE_PRINCIPAL
+		connDiag, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, actionErr)
+		return connDiag, trace.Wrap(err)
+	}
+
+	if kubeErr != nil {
+		traceType := types.ConnectionDiagnosticTrace_UNKNOWN_ERROR
+		message := fmt.Sprintf("Unknown error. %v", actionErr)
+		connDiag, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, actionErr)
+		return connDiag, trace.Wrap(err)
+	}
+
+	message = "You are authorized to access this Kubernetes Cluster."
+	traceType = types.ConnectionDiagnosticTrace_RBAC_KUBE
+
+	connDiag, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return connDiag, nil
+}
+
+func (s KubeConnectionTester) handleUserGenCertsErr(ctx context.Context, connectionDiagnosticID string, actionErr error) (types.ConnectionDiagnostic, error) {
+	if trace.IsBadParameter(actionErr) {
+		message := "Failed to connect to kubernetes cluster. Ensure the cluster is registered."
+		traceType := types.ConnectionDiagnosticTrace_CONNECTIVITY
+		return s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, actionErr)
+	} else if actionErr != nil {
```

Comment:
nit pick 
```suggestion
	}
	
	if actionErr != nil {
```

You're returning in your first branch so no need to do an `else`